### DISCUSSION TITLE
Is a tcmenuDesigner aarch64 deb build possible for NOMone Desktop Ubuntu Linux proot on Android mobile?

### DISCUSSION THREAD

**Discussion by AtmosphEng** [TcMenu/tcMenu#497] [API & Remote]
I would like to use my Android mobile as a remote using tcMenuDesigner. I have Ubuntu 22.04 running on my mobile via NOMone, which is available on the PlayStore. App installation on MODone is only available with aarch64 deb builds. Is a aarch64 build of the current tcMenuDesigner possible or feasible?
Thanks for your fantastic work on tcmenu and I've become a regular sponsor. Please keep on providing the community with your very high quality software.

**Comment by davetcc** [comment]
I think that as long as a packaged OpenJDK is available and built (along with OpenJFX) it should be possible at least. 

There are instructions how to build it for any platform in the tcMenuGenerator/scripts folder, but we're in the middle of making that easier. 

Soon it would be possible to run nightly builds just by pulling them from here and having OpenJDK and OpenJFX installed. I suspect if anything would be tricky, it would be finding a built version of OpenJFX for that version of Linux.

**Reply by davetcc** [reply]
sorry for the delay it's been a busy couple of weeks with a bit of reorganisation, see the new discussion if youre interested in the details.

**Reply by AtmosphEng** [reply]
Thanks for the new release. I've upgraded and found no issues so far.
I've asked the developers of NOMone Desktop (Ubuntu on Android) if they support OpenJDK and OpenJFX, but it appears not. I'll update here if I hear any changes.
As an alternative, do you have any timescale you can share for a tcmenu build of embedControl for Android?

**Reply by AtmosphEng** [reply]
Post to NOMone Desktop Forum:
[Can OpenJDK and OpenJFX be installed on NOMone Desktop Ubuntu?

I want to install a deb package on NOMone Desktop on my Android Mobile. I understand it need to be for aarch64. The developer of the deb package says it needs OpenJDK and OpenJFX. Is this possible on NOMone Desktop? Thanks.

Reply from NOMone Desktop Forum:
Yes, I'm sure you can install OpenJDK but I didn't try to install OpenJFX before. It would be great if you give it a try and let us know how it goes. I think should work properly similar to OpenJDK isA : )

To davetcc:
Could you provide an outline procedure for me to try and check if tcmenuDesigner and/or embedControl is feasible on Android Linux via NOMone Desktop, or some other way to progress getting embedControl to work on Android? Thanks.

**Reply by davetcc** [reply]
The outline of building everything is here: it may not be 100% up to date but it should get you started, and you could raise any issues here.

https://github.com/TcMenu/tcMenu/blob/main/tcMenuGenerator/scripts/packager-all-platforms.md

The plan is to get the maven output from github actions and create a kind of nightly build somewhere. It would remove the need for the maven steps.

**Reply by davetcc** [reply]
As far as OpenJFX goes, it must be built for that environment, but as it is already built for both Android and for Linux Arm64 it is possible one already exists.

**Reply by AtmosphEng** [reply]
Thanks for the pointers. Currently I have the following preliminary setup:
0) AMD64 test machine
1) Ubuntu 24.04.1 LTS
2) openjdk version "21.0.4" 2024-07-16
3) javac 21.0.4
4) Apache Maven 3.8.7

Local cloned repository via
git clone https://github.com/TcMenu/tcMenu.git
git checkout main  (should this be main of some other branch?)

I've tried a  "mvn clean install", in each of the following git locally cloned directories:
tcMenu/tcMenuGenerator and tcMenu/embedCONTROLCore,
but both Maven builds fail with errors beginning with: error: invalid target release: 22 -> [Help 1]
[Help 1] is a link to MojoExecutionException and says this exception is not caused by the Maven core but is caused by a plugin(s).

Some questions please:
a) Any ideas on how to progress from here?
b) Will the final compliation for the Android phone using aarch64 be possibly cross-compiled from my AMD64 machine or does it need to be done on the Android phone target itself?

Thanks.

**Reply by davetcc** [reply]
Quick feedback: You'll need to get java version 22 or 23. We always use the most recent version available for designer. The API will compile with 11 upwards though.

**Reply by AtmosphEng** [reply]
Thanks for the feedback. I've updated my versions as follows:

LATEST VERSIONS:
openjdk 23+37 from John Neffenger (jgneff) installed
openjfx 23+29 from John Neffenger (jgneff) installed
javac 23
java version "23" 2024-09-17
Apache Maven 3.9.9
libmockito-java (2.23.0-2)

RESULTS 1:
cd ~/tcMenu/tcMenuGenerator
mvn clean install
Mockito cannot mock this class: class com.thecoderscorner.menu.editorui.uimodel.CurrentProjectEditorUIImpl.
If you're not sure why you're getting this error, please open an issue on GitHub.
[INFO] 
[ERROR] Tests run: 118, Failures: 0, Errors: 25, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  54.860 s
[INFO] Finished at: 2024-10-02T17:52:05+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.4.0:test (default-test) on project menuEditorUI: 
[ERROR] 

RESULTS 2:
cd ~/tcMenu/embedCONTROLCore
mvn clean install
[WARNING] 100 warnings
1x [ERROR] Failed to execute goal org.apache.maven.plugins:maven-gpg-plugin:3.2.5:sign (sign-artifacts) on project embedCONTROLCore: Exit code: 2 -> [Help 1]

CONCLUSION:
Some progress made. Could you provide some fresh pointers please.

**Reply by davetcc** [reply]
The following command should do it:

mvn install -DskipTests -Dgpg.skip

**Comment by davetcc** [comment]
That is odd that the tests fail. As they are running locally on the CI on Java 22 and soon to be 23. 

Could you try to add -DskipTests to the maven build command and try again.

**Reply by AtmosphEng** [reply]
Yes, that make a difference.
I've tried building both tcMenuGenerator and embedCONTROLCore with:
mvn install -DskipTests -Dgpg.skip
and now both report:
[INFO] BUILD SUCCESS

**Reply by AtmosphEng** [reply]
Some more progress, thanks.
I've made a gpgkey and built the api, then I ran the "smoke test":
_Packaging - Check the archive worked and versioning is right
Run and smoke test, ensure you are in the tcMenuGenerator/target/jfx/app directory_
This ran the tcMenuDesigner app ok, but I had to move the app to an extended hdmi display for the drop-down menus to become visible. I successfully created and made an embedCONTROL connection to my esp32-s3 target.

But I had errors at the package stage.
cd ~/tcMenu/tcMenuGenerator/target
_jpackage -n tcMenuDesigner -p jfx/deps --input jfx/app --icon ./classes/img/menu-icon.png --verbose --license-file ../../LICENSE --linux-app-category Development --linux-menu-group "Development;Utility;" --java-options "-Dprism.lcdtext=false --enable-preview -Djava.library.path=$APPDIR/lin" --app-version 4.3.1  --add-modules "jdk.crypto.cryptoki" -m com.thecoderscorner.tcmenu.menuEditorUI/com.thecoderscorner.menu.editorui.cli.TcMenuDesignerCmd_
This produced errors as follows:
ERRORS:
[21:05:20.359] Output:
    Error: java.io.IOException: Cannot run program "objcopy": error=2, No such file or directory
[21:05:20.360] Returned: 1
[21:05:20.363] jdk.jpackage.internal.PackagerException: jlink failed with: Error: java.io.IOException: Cannot run program "objcopy": error=2, No such file or directory

Any ideas to try and fix this please.

**Reply by davetcc** [reply]
Probably best to read the jpackage instructions for Linux. You need to have all the native packaging tools installed for the platform. The should be documented on that page.

**Reply by AtmosphEng** [reply]
Thanks. Yesterday I installed fakeroot and today I added binutils (for objcopy etc).
After this, I got:
[11:44:13.923] Succeeded in building DEB Bundle package

This then allowed a successful install and run of tcmenudesigner (4.3.1).
Now I need to see if this all works on Android proot of Linux Ubuntu via NOMone Desktop.

**Reply by AtmosphEng** [reply]
SUCCESS: tcMenuDesigner > embedCONTROL works on my Android mobile phone via NOMone Desktop Ubuntu Linux.
Your "smoke test" and the deb file package both ran ok.
My mobile phone spec is (about 7 years old): Samsung Galaxy A7 (2017),  Android 8.0.0,  32 GB internal storage, 3 GB RAM, 1 TB SD card, 64-bit Arm.
Thanks a lot for your help on this !

**Comment by AtmosphEng** [comment]
PROBLEM USING tcMenuDesigner>embedCONTROL with a Serial connection.
I found that running the latest Windows and Linux versions of tcMenuDesigner runs esp32 wifi connections fine, but they cannot run embedCONTROL over Serial. In all cases, when a serial connection attempt is made, no spawned tcMenu window appears, and nothing else seems to happen.
I then went back and tried an older version from setupEmbedCONTROL-4.1.1.exe, and this does run serial connections ok.

To follow this up, I tried a local build of embedCONTROL on Ubuntu.
Following the embedCONTROL build instructions at the bottom of: https://github.com/TcMenu/tcMenu/blob/main/tcMenuGenerator/scripts/packager-all-platforms.md, 
I found that running the final package:
**java --module-path ../deps "-Dprism.lcdtext=false" -m com.thecoderscorner.tcmenu.embedcontrolfx/com.thecoderscorner.embedcontrol.jfxapp.EmbedControlApp**
resulted in the following error:
_Error occurred during initialization of boot layer
java.lang.module.FindException: Module com.thecoderscorner.tcmenu.embedcontrolfx not found_

Could you please take a look running embedCONTROL over serial connections and advise. Thanks.

**Reply by AtmosphEng** [reply]
FYI - I've just made an additional sponsorship payment to you as thanks for your support on embedCONTROL wifi and serial connections.

**Reply by AtmosphEng** [reply]
I also tried the debian package for embedCONTROL 4.1.1, and it did not connect via serial (/dev/ttyACM0) unless I ran embedCONTROL as super user e.g. sudo /opt/embedcontrol/bin/embedCONTROL. With sudo, the embedCONTROL serial connection was successful.

**Reply by davetcc** [reply]
Thanks a lot for the sponsorship, it really helps us to keep paying for the certificates and licences needed. As it happens, we're in the middle of a full retest of every board, and I'll make sure the boards that have serial remote connections are tested carefully.

**Reply by AtmosphEng** [reply]
You're welcome. 
In the meantime is is possible for you to provide any instructions to building (possibly older) versions of standalone embedCONTROL, which I can try building on aarch64? 
From above, I got this error:
java --module-path ../deps "-Dprism.lcdtext=false" -m com.thecoderscorner.tcmenu.embedcontrolfx/com.thecoderscorner.embedcontrol.jfxapp.EmbedControlApp
resulted in the following error:
Error occurred during initialization of boot layer
java.lang.module.FindException: Module com.thecoderscorner.tcmenu.embedcontrolfx not found

**Reply by davetcc** [reply]
All the releases are tagged, so you should just be able to use git to checkout any of the older versions, for example let's say you wanted to check out `4.1.1` I would think something like `git checkout tags/4.1.1` should work from the repository directory. Then the `tcMenuGenerator/scripts` directory will still have the instructions for building the standalone embed control.

**Reply by AtmosphEng** [reply]
Thanks. My Git skills are improving, so I'll happily give this a try.

**Comment by davetcc** [comment]
So on the other question, what are intending to do with embed control.

We've had a long discussion internally and realised that by trying to make a one size fits all UI for mobile with our own form designer, we've made it impossible to deliver with the number of people presently working of the project, which boils down to three people's voluntary efforts. @vzahradnik  @Ales-Svoboda  and I have been discussing what we think is possible, and the plan is to also open-source the Flutter/Dart components that we have (including the "Flutter Embed Control" app as a starting point) soon. For most people, this would be the easiest way to get something onto a mobile device. Our plans for this are twofold:

1. A Java API along with the components that make up embed control. These will allow you to do what you're doing now, but will also be a bit more flexible, so if you wanted a custom designed page instead of the Auto UI this would be much easier to achieve than now.
2. A Dart API along with the flutter components that provide an Auto UI somewhat similar to the current Java offering. Again, the components will be provided in a way that you can easily extend them to create your own custom UI too. With the flutter UI it would be possible to create an apk for Android, an app file for iPhone and a Windows store package too.

The license for the dart API will likely be Apache, but the flutter Auto UI starter app will most likely be more restrictive, including a clause for commercial use and another that it is never deployed as a general-purpose app to a Store. I.E. that it is only ever specialised to allow connections to your own hardware. The Flutter "starting point app" would be about the same as the Java embed control UI, but it would be deployable to Android as an apk and Apple devices as an app bundle.

We're also looking to integrate much more closely with MQQT and existing home ecosystems, probably the open source ones first. You'll see the direction of travel for this in the issue tracking. I mean this is really open to a lot of discussion and we'd be really glad to get as much feedback as possible, and even more happy for others to join the effort and help us deliver these things.

**Reply by AtmosphEng** [reply]
Many thanks for the detailed update. My skills are mainly C and C++, but I might be able to help with a bit of direction.

**Reply by vzahradnik** [reply]
I will just add two things:
- Java API can be used also on Android because of the interoperability with Kotlin. If I would build a custom app, I would use Java API as the most stable
- We have a public Discord server where you can join as well. At the moment we are brainstorming the way to go for the TcMenu long-term

As a side note, more hands on the project are always helpful. If you'd ever consider contributing also by writing some code, we will gladly guide you. There is a lot of things to work on.

**Reply by AtmosphEng** [reply]
After sleeping on your kind offer to guide me on coding, I have to so sorry, but no thanks. I've realised that I've still got lots of my own project ideas to concentrate on. My background is also in electrical, electronics, and education, so maybe I'll do some Youtube videos showing my use of tcMenu on my main monorail train project etc..

**Reply by vzahradnik** [reply]
Videos sound cool. We welcome contributions like this since TcMenu is still rather unknown in the wild.

**Comment by AtmosphEng** [comment]
FYI:
I've been updating the NOMone Desktop support site with some remote control build details on aarch64.:
https://desktop.nomone.com/forum/en/index.php

NOMone asked me for my sample test code, so I published this:
https://github.com/AtmosphEng/PublicTcmSimple.git

**Comment by vzahradnik** [comment]
@AtmosphEng here is our Discord. If you'd like to help with brainstorming or just hang out.

https://discord.gg/MHuRz6RuSx

**Comment by davetcc** [comment]
I've found the issue with serial connections, it unfortunately will need a release, but we've done all the testing and one is coming very soon: https://github.com/TcMenu/tcMenu/issues/531

If you really wanted not to wait, you could apply the same changes that will be in #531 as they are trivial.

**Reply by davetcc** [reply]
In fact the required change got committed on the version number bump instead of the above issue. It was in `embedCONTROLCore/pom.xml`:

        <dependency>
            <groupId>com.thecoderscorner.tcmenu</groupId>
            <artifactId>tcMenuJavaAPI</artifactId>
            <version>4.2.1</version>   <<<<--- change this to ${project.version} and rebuild
        </dependency>

**Reply by AtmosphEng** [reply]
Cool. Thanks for following that up.

**Comment by davetcc** [comment]
This post is still very popular. Something that was missed when this was initially replied to is that a starter OpenJFX App, along with the original full version of Embed Control is hosted here: https://github.com/TcMenu/tcmenu-examples-starters/, it's a bit out of date but I purposely keep the starter repo on older deps by default as they are sometimes used in places where a newer Java may not be available.
